Code Analysis: Performance considerations

MSVCR 2013 is throwing some warnings when compiling libsass code via node-sass. Some of the warnings are related to the compiler (cl.exe) switches, but mostly its about the code-analysis warnings, which might help improving libsass code base (code-cleanup, best practices etc.). I guess it's worth checking out.

Following is the log (can also be found on Appveyor-CI for testing node-sass on WindOS: https://ci.appveyor.com/project/sass/node-sass/build/31/job/khsoq7bnmnajaqcf -- lines 35 to 84)

> C:\projects\node-sass>node "C:\Program Files (x86)\nodejs\node_modules\npm\bin\node-gyp-bin\....\node_modules\node-gyp\bin\node-gyp.js" rebuild  
> Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
> cl : Command line warning D9025: overriding '/GR-' with '/GR' [C:\projects\node-sass\build\binding.vcxproj]
> binding.cpp
> C:\Users\appveyor.node-gyp\0.10.32\deps\v8\include\v8.h(179): warning C4506: no definition for inline function 'v8::Persistentv8::Object v8::Persistentv8::Object::New(v8::Handlev8::Object)' [C:\projects\node-sass\build\binding.vcxproj]
> sass_context_wrapper.cpp
> C:\Users\appveyor.node-gyp\0.10.32\deps\v8\include\v8.h(179): warning C4506: no definition for inline function 'v8::Persistentv8::Object v8::Persistentv8::Object::New(v8::Handlev8::Object)' [C:\projects\node-sass\build\binding.vcxproj]
> ast.cpp
> base64vlq.cpp
> bind.cpp
> c:\projects\node-sass\libsass\environment.hpp(36): warning C4800: 'unsigned int' : forcing value to bool 'true' or 'false' (performance warning) [C:\projects\node-sass\build\binding.vcxproj]
>         c:\projects\node-sass\libsass\environment.hpp(36) : while compiling class template member function 'bool Sass::Environment<Sass::AST_Node *>::current_frame_has(const std::string) const'
>         ..\libsass\bind.cpp(44) : see reference to function template instantiation 'bool Sass::Environment<Sass::AST_Node *>::current_frame_has(const std::string) const' being compiled
>         ..\libsass\bind.cpp(44) : see reference to class template instantiation 'Sass::Environment<Sass::AST_Node *>' being compiled
> constants.cpp
> context.cpp
> contextualize.cpp
> copy_c_str.cpp
> error_handling.cpp
> eval.cpp
> ..\libsass\eval.cpp(854): warning C4800: 'int' : forcing value to bool 'true' or 'false' (performance warning) [C:\projects\node-sass\build\binding.vcxproj]
> expand.cpp
> extend.cpp
> file.cpp
> C:\Program Files (x86)\Windows Kits\8.1\Include\um\winbase.h(850): warning C4005: 'FS_CASE_SENSITIVE' : macro redefinition [C:\projects\node-sass\build\binding.vcxproj]
>       ..\libsass\file.cpp(7) : see previous definition of 'FS_CASE_SENSITIVE'
> ..\libsass\file.cpp(229): warning C4101: 'st' : unreferenced local variable [C:\projects\node-sass\build\binding.vcxproj]
> functions.cpp
> inspect.cpp
> node.cpp
> output_compressed.cpp
> output_nested.cpp
> parser.cpp
> ..\libsass\parser.cpp(274): warning C4800: 'const char *' : forcing value to bool 'true' or 'false' (performance warning) [C:\projects\node-sass\build\binding.vcxproj]
> ..\libsass\parser.cpp(275): warning C4800: 'const char *' : forcing value to bool 'true' or 'false' (performance warning) [C:\projects\node-sass\build\binding.vcxproj]
> ..\libsass\parser.cpp(1065): warning C4101: 'err' : unreferenced local variable [C:\projects\node-sass\build\binding.vcxproj]
> Compiling...
> prelexer.cpp
> remove_placeholders.cpp
> sass.cpp
> sass2scss.cpp
> sass_interface.cpp
> sass_util.cpp
> source_map.cpp
> to_c.cpp
> to_string.cpp
> units.cpp
> utf8_string.cpp
> util.cpp
>       Creating library C:\projects\node-sass\build\Release\binding.lib and object C:\projects\node-sass\build\Release\binding.exp
> Generating code
> Finished generating code
> binding.vcxproj -> C:\projects\node-sass\build\Release\binding.node
> Installed in `C:\projects\node-sass\bin\win32-ia32-v8-3.14\binding.node

--- a/ast.hpp
+++ b/ast.hpp
@@ -658,7 +658,7 @@ namespace Sass {
           if (!(*(elements()[i]) == *(l[i]))) return false;
         return true;
       }
-      catch (std::bad_cast& bc)
+      catch (std::bad_cast&)
       {
         return false;
       }
@@ -706,7 +706,7 @@ namespace Sass {
           if (!(*at(key) == *m.at(key))) return false;
         return true;
       }
-      catch (std::bad_cast& bc)
+      catch (std::bad_cast&)
       {
         return false;
       }
@@ -815,7 +815,7 @@ namespace Sass {
         Variable& e = dynamic_cast<Variable&>(rhs);
         return e && name() == e.name();
       }
-      catch (std::bad_cast& bc)
+      catch (std::bad_cast&)
       {
         return false;
       }
@@ -969,7 +969,7 @@ namespace Sass {
         e.normalize(find_convertible_unit());
         return unit() == e.unit() && value() == e.value();
       }
-      catch (std::bad_cast& bc)
+      catch (std::bad_cast&)
       {
         return false;
       }
@@ -1008,7 +1008,7 @@ namespace Sass {
         Color& c = (dynamic_cast<Color&>(rhs));
         return c && r() == c.r() && g() == c.g() && b() == c.b() && a() == c.a();
       }
-      catch (std::bad_cast& bc)
+      catch (std::bad_cast&)
       {
         return false;
       }
@@ -1044,7 +1044,7 @@ namespace Sass {
         Boolean& e = dynamic_cast<Boolean&>(rhs);
         return e && value() == e.value();
       }
-      catch (std::bad_cast& bc)
+      catch (std::bad_cast&)
       {
         return false;
       }
@@ -1120,7 +1120,7 @@ namespace Sass {
         String_Constant& e = dynamic_cast<String_Constant&>(rhs);
         return e && unquoted_ == e.unquoted_;
       }
-      catch (std::bad_cast& bc)
+      catch (std::bad_cast&)
       {
         return false;
       }

--- a/environment.hpp
+++ b/environment.hpp
@@ -33,7 +33,7 @@ namespace Sass {
     }
 
     bool current_frame_has(const string key) const
-    { return current_frame_.count(key); }
+    { return !!current_frame_.count(key); }
 
     Environment* grandparent() const
     {

--- a/eval.cpp
+++ b/eval.cpp
@@ -853,7 +853,7 @@ namespace Sass {
     Expression* e = 0;
     switch (v.unknown.tag) {
       case SASS_BOOLEAN: {
-        e = new (ctx.mem) Boolean(path, position, v.boolean.value);
+        e = new (ctx.mem) Boolean(path, position, !!v.boolean.value);
       } break;
       case SASS_NUMBER: {
         e = new (ctx.mem) Number(path, position, v.number.value, v.number.unit);

--- a/file.cpp
+++ b/file.cpp
@@ -2,14 +2,6 @@
 #define S_ISDIR(mode) (((mode) & S_IFMT) == S_IFDIR)
 #endif
 
-#ifndef FS_CASE_SENSITIVE
-#ifdef _WIN32
-#define FS_CASE_SENSITIVE 0
-#else
-#define FS_CASE_SENSITIVE 1
-#endif
-#endif
-
 #include <iostream>
 #include <fstream>
 #include <cctype>
@@ -22,6 +14,14 @@
 
 #ifdef _WIN32
 #include <windows.h>
+#endif
+
+#ifndef FS_CASE_SENSITIVE
+#ifdef _WIN32
+#define FS_CASE_SENSITIVE 0
+#else
+#define FS_CASE_SENSITIVE 1
+#endif
 #endif
 
 namespace Sass {
@@ -226,8 +226,6 @@ namespace Sass {
 
     char* read_file(string path)
     {
-      struct stat st;
-
 #ifdef _WIN32
       BYTE* pBuffer;
       DWORD dwBytes;
@@ -244,6 +242,7 @@ namespace Sass {
       // just convert from unsigned char*
       char* contents = (char*) pBuffer;
 #else
+      struct stat st;
       if (stat(path.c_str(), &st) == -1 || S_ISDIR(st.st_mode)) return 0;
       ifstream file(path.c_str(), ios::in | ios::binary | ios::ate);
       char* contents = 0;

--- a/functions.cpp
+++ b/functions.cpp
@@ -1192,7 +1192,7 @@ namespace Sass {
       Expression* v = ARG("$key", Expression);
       try {
         return m->at(v);
-      } catch (const std::out_of_range& oor) {
+      } catch (const std::out_of_range&) {
         return new (ctx.mem) Null(path, position);
       }
     }

--- a/parser.cpp
+++ b/parser.cpp
@@ -271,8 +271,8 @@ namespace Sass {
     if (!lex< exactly<':'> >()) error("expected ':' after " + name + " in assignment statement");
     Expression* val = parse_list();
     val->is_delayed(false);
-    bool is_guarded = lex< default_flag >();
-    bool is_global = lex< global_flag >();
+    bool is_guarded = lex< default_flag >() != NULL;
+    bool is_global = lex< global_flag >() != NULL;
     Assignment* var = new (ctx.mem) Assignment(path, var_source_position, name, val, is_guarded, is_global);
     return var;
   }
@@ -1070,7 +1070,7 @@ namespace Sass {
         *args << arg;
         return result;
       }
-      catch (Error& err) {
+      catch (Error&) {
         // back up so we can try again
         position = here;
         source_position = here_p;

--- a/sass.cpp
+++ b/sass.cpp
@@ -169,7 +169,7 @@ extern "C" {
   // make_sass_* may allocated additional memory
   void free_sass_value(const union Sass_Value val) {
 
-    int i;
+    size_t i;
     switch(val.unknown.tag) {
         // case SASS_NULL: {
         // }   break;

